File: spiral.c
#include <stdio.h>
#include <stdlib.h>

void spirallyTraverse(int** mat, int r, int c) {
    int top = 0, bottom = r - 1, left = 0, right = c - 1;

    printf("Spiral Traversal: ");
    while (top <= bottom && left <= right) {
        for (int i = left; i <= right; i++)
            printf("%d ", mat[top][i]);
        top++;
     
        for (int i = top; i <= bottom; i++)
            printf("%d ", mat[i][right]);
        right--;

        if (top <= bottom) {
            for (int i = right; i >= left; i--)
                printf("%d ", mat[bottom][i]);
            bottom--;
        }

        if (left <= right) {
            for (int i = bottom; i >= top; i--)
                printf("%d ", mat[i][left]);
            left++;
        }
    }
    printf("\n");
}

int main() {
    int r, c;
    printf("Enter number of rows: ");
    scanf("%d", &r);
    printf("Enter number of columns: ");
    scanf("%d", &c);

    int** mat = (int**)malloc(r * sizeof(int*));
    for (int i = 0; i < r; i++) {
        mat[i] = (int*)malloc(c * sizeof(int));
    }

    printf("Enter the matrix elements:\n");
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            scanf("%d", &mat[i][j]);
        }
    }

    spirallyTraverse(mat, r, c);

    for (int i = 0; i < r; i++) {
        free(mat[i]);
    }
    free(mat);

    return 0;
}

File: spiral.py
rows = int(input("Enter the number of rows: "))
cols = int(input("Enter the number of columns: "))
matrix = []
top = 0; bottom = rows - 1
left = 0; right = cols - 1

print("Input matrix:")
for i in range(rows):
    row = []
    for j in range(cols):
        value = int(input(f"Element [{i}][{j}]: "))
while (top <= bottom and left <= right):

    for i in range(left, right + 1):
        print(matrix[top][i], end=' ')
    top += 1

    for i in range(top, bottom + 1):
        print(matrix[i][right], end=' ')
    right -= 1

    if top <= bottom:
        for i in range(right, left - 1, -1):
            print(matrix[bottom][i], end=' ')
        bottom -= 1

    if left <= right:
        for i in range(bottom, top - 1, -1):
            print(matrix[i][left], end=' ')
        left += 1ght:
        for i in range(bottom, top - 1, -1):
            print(matrix[i][left], end=' ')
        left += 1